cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,251 @@
+package ext
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/bash"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/powershell"
+)
+
+// Generic action function that may return an error
+type ActionFn func() error
+
+// The type of command hooks. Supported values are 'pre' and 'post'
+type HookType string
+
+const (
+	// Executes pre-command hooks
+	HookTypePre HookType = "pre"
+	// Execute post-command hooks
+	HookTypePost HookType = "post"
+)
+
+// CommandHooks enable support to invoke integration scripts before & after commands
+// Scripts can be invoked at the project or service level or
+type CommandHooks struct {
+	commandRunner exec.CommandRunner
+	console       input.Console
+	cwd           string
+	interactive   bool
+	scripts       map[string]*ScriptConfig
+	envVars       []string
+}
+
+// NewCommandHooks creates a new instance of CommandHooks
+func NewCommandHooks(
+	commandRunner exec.CommandRunner,
+	console input.Console,
+	scripts map[string]*ScriptConfig,
+	cwd string,
+	envVars []string,
+	interactive bool,
+) *CommandHooks {
+	return &CommandHooks{
+		commandRunner: commandRunner,
+		console:       console,
+		cwd:           cwd,
+		interactive:   interactive,
+		scripts:       scripts,
+		envVars:       envVars,
+	}
+}
+
+// Invokes an action run runs any registered pre or post script hooks for the specified command.
+func (h *CommandHooks) InvokeAction(ctx context.Context, commandName string, actionFn ActionFn) error {
```

Comment:
The term `Action` here is not to be confused with our top level `Action` executed from cobra commands but refers to any simple execution function that takes a go context as input and returns an error.

Reply:
This is one of the more interesting functions that executes all the pre/post command hooks around any specified action function.

This is the function used for our sub deploy commands like `package`.